Text based Circuit Rendering Module

gadhvirushiraj opened the pull request:
## Text based Circuit Rendering Module 
####  GSOC'24 Project - QuTiP: Enhanced Quantum Circuit Diagrams | Phase#2 Draft PR

#### Major Task Outline

- [x] Single Qubit Gate Rendering
- [x] Multi-Qubit Gate Rendering
- [x] Custom Gate Rendering
- [x] Measurement Rendering 
- [ ] Customization Options
- [ ] Jupyter Notebook Tutorials

BoxiLi commented on tests/test_textrenderer.py:
```diff
@@ -0,0 +1,83 @@
+            parts, _ = tr._draw_measurement_gate(gate)
+
+        len_parts = [len(part) for part in parts]
+        assert (
```
Add a comment that this is to test all the `len_parts` are the same.

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,506 @@
+    "TextRenderer",
+]
+
+
+class TextRenderer:
+    """
+    A class to render a quantum circuit in text format.
+    """
+
+    def __init__(self, qc: QubitCircuit):
+
+        self._qc = qc
+        self._qwires = qc.N
```
A design question, technically speaking, the Renderer class should be independent of the quantum circuit provided.

Therefore, `qc` should better be an input argument for some `canvas_plot` for `draw` methods. Not as part of the initialization of the renderer class.

BoxiLi commented:
There is one thing that I forgot last time, the public API should be added to the apidoc https://github.com/qutip/qutip-qip/tree/master/doc/source/apidoc

BoxiLi commented on src/qutip_qip/circuit/mat_renderer.py:
```diff
@@ -15,130 +15,16 @@
-
-    gate_pad : float, optional
-        The padding between the gate and the gate label. The default is 0.05.
-
-    label_pad : float, optional
-        The padding between the wire label and the wire. The default is 0.1.
-
-    fig_height : float, optional
-        The height of the figure. The default is None.
-
-    fig_width : float, optional
-        The width of the figure. The default is None.
-
-    bulge : Union[str, bool], optional
-        The bulge style of the gate. Renders non-bulge gates if False. The default is True.
-
-    align_layer : bool, optional
-        Align the layers of the gates across different wires. The default is False.
-
-    theme : Optional[Union[str, Dict]], optional
-        The color theme of the circuit. The default is "qutip".
-        The available themes are 'qutip', 'light', and 'modern'.
-
-    title : Optional[str], optional
-        The title of the circuit. The default is None.
-
-    bgcolor : Optional[str], optional
-        The background color of the circuit. The default is None.
-
-    color : Optional[str], optional
-        Controls color of acsent elements (eg. cross sign in the target node)
-        and set as deafult color of gate-label. Can be overwritten by gate specific color.
-        The default is None.
-
-    wire_label : Optional[List], optional
-        The labels of the wires. The default is None.
-
-    wire_color : Optional[str], optional
-        The color of the wires. The default is None.
-    """
```
How do I know which are available for the Textrender and which are for the Matrenderer? Is there a way to document it or make it clearer?

BoxiLi replied:
These options should probably be summarised or somehow demonstrated to the user either in the user guide or in the doc string referred to in the user guide. Especially how they can be defined via the `circuit.draw` interface, as dict? Or as class?

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+                self._render_strs["mid_frame"][wire] += mid_bar_conn
+                self._render_strs["bot_frame"][wire] += bar_conn
+
+    def layout(self):
```
What is the difference between `layout` and `print_circuit`?

gadhvirushiraj replied:
The `layout` function iterates through the gates and updates the `_render_strs` list of strings, while `print_circuit` handles displaying the `_render_strs` in the correct order, either in the terminal or within an IPython environment.

BoxiLi reviewed (commented):
This is just a first quick review. A more detailed one will follow.

You can also add the test for matplotlib here, should be fine. There should also be one test for the circuit.draw() interface, making sure that the API is working properly.

The tex renderer should also be renamed to match the convention.

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+
+    def _add_wire_labels(self):
+        """
+        Add wire labels to the circuit
+        """
+
+        default_labels = [f"q{i}" for i in range(self._qwires)] + [
+            f"c{i}" for i in range(self._cwires)
+        ]
+        for i, label in enumerate(default_labels):
+            self._render_strs["mid_frame"][i] = (
+                f" {label} :" + self._render_strs["mid_frame"][i]
+            )
+            self._render_strs["top_frame"][i] = " " * (len(f" {label} :") + 2)
+            self._render_strs["bot_frame"][i] = " " * (len(f" {label} :") + 2)
+
+        self._layer_list = {
+            i: [len(self._render_strs["mid_frame"][i])]
+            for i in range(self._qwires + self._cwires)
+        }
```
What does `_layer_list` actually save? I assume `_layer_list[i]` is a list of length for each element for a wire? Maybe document it somewhere.i

gadhvirushiraj replied:
Dict was used to address the classical wire issue with measurement joint, though it was handled differently. So, its a list.

Currently, it shifts the entire wire by a space for `q10` because of extra digit.  Will make change to add padding to make all label len same.

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+            for i in range(self._qwires + self._cwires)
+        }
+
+    def _extend_line(self):
+        """
+        Extend all the wires to the same length
+        """
+
+        wire_len = [sum(wire) for wire in self._layer_list.values()]
+        max_len = max(wire_len) + self.style.end_wire_ext
+
+        for i, length in enumerate(wire_len):
+            if length < max_len:
+                diff = max_len - length
+                if i < self._qwires:
+                    self._render_strs["mid_frame"][i] += "─" * diff
+                else:
+                    self._render_strs["mid_frame"][i] += "═" * diff
```
What is the difference between `_extend_line` and `_adjust_layer`. Why does `_extend_line` not add space to the top and bottom? I feel like `_extend_line` can be replaced by `adjust_layer` by a proper definition of `xskip`

gadhvirushiraj replied:
- Since we're dealing with strings and the gates vary in width, `_adjust_layer` adds padding to maintain consistent layer alignment across the different wires. 

- `extend_wire` is only used at the end of circuit, after all gates are added to extend wire to certain extra length

gadhvirushiraj replied:
It's a nice idea to reuse the `_adjust_layer` instead of `_extend_wire`, will give it a try.

gadhvirushiraj replied:
Update: Removed `_extend_line`

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+            The width of the gate.
+        """
+
+        lid_seg = "─" * (ceil(self.style.gate_pad) * 2 + len(gate_name))
```
From the default Style, `gate_pad=0.05`, however here I believe we need an integer here? `ceil` is a bit weird. It will cause weird behaviour. I think we should just have a different default value. It seems that it would be nice to summarize the default value somewhere. This is related to the other comment of mine yesterday. Maybe have a separate default style for each renderer?

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+        if gate.controls:
+            mid_index = len(bot_frame) // 2
+            if gate.controls[0] < gate.targets[0]:
+                bot_frame = (
+                    bot_frame[:mid_index] + "┬" + bot_frame[mid_index + 1 :]
+                )
```
I think you  can just use `bot_frame[:mid_index]="┬"`

BoxiLi commented on src/qutip_qip/circuit/base_renderer.py:
```diff
@@ -0,0 +1,212 @@
+            The maximum xskip value needed to reach the specified layer.
+        """
+
+        if self.style.align_layer:
+            wire_list = list(range(self._qwires))
+
+        xskip = []
+        for wire in wire_list:
+            xskip.append(sum(self._layer_list[wire][:layer]))
+
+        return max(xskip)
```
Does `aligh_layer` also work for the text renderer?

BoxiLi replied:
I guess text rendering is by default aligned.

gadhvirushiraj replied:
Yes, it does perfectly 😃

gadhvirushiraj replied:
> I guess text rendering is by default aligned.

It’s not enabled by default—we still need to set the `align_layer=True` argument. This is essential because we’re continuing to use the layer concept to store gates at different horizontal positions, aiming for maximum compactness. The only change is that the minimum shift in the case of text is now 1 unit, rather than a floating-point value.

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+        mid_bar_conn = "─" * (width // 2) + "║" + "─" * (width // 2)
+        mid_bar_classical_conn = "═" * (width // 2) + "║" + "═" * (width // 2)
+        classical_conn = "═" * (width // 2) + "╩" + "═" * (width // 2)
+        self._adjust_layer(wire_list, xskip)
```
There is an `_adjust_layer` for all `_update_*` methods except for `_update_singleq`, why?

gadhvirushiraj replied:
Padding via `_adjust_layer` is primarily necessary when a gate interacts with multiple wires, as it ensures consistent padding across all of them.

However, while testing with align_layer=True, I discovered an edge case where additional padding is needed. I'll likely implement this adjustment after each gate is added.

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+            layer = max(len(self._layer_list[i]) for i in wire_list)
+            xskip = self._get_xskip(wire_list, layer)
+            self._manage_layers(width, wire_list, layer, xskip)
+
+            if isinstance(gate, Measurement):
+                self._adjust_layer(wire_list, xskip)
+                self._update_singleq(gate.targets, parts)
+                self._update_cbridge(gate, wire_list, xskip, width)
+            elif len(gate.targets) == 1 and gate.controls is None:
+                self._update_singleq(wire_list, parts)
+            else:
+                sorted_targets = sorted(gate.targets)
+                self._update_target_multiq(
```
Is it possible to put one `self._adjust_layer(wire_list, xskip)` in line 419, i.e., before all the `_update_*` methods. And remove all of those inside the `_update_*` methods?

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+            elif len(gate.targets) == 1 and gate.controls is None:
+                wire_list = gate.targets
+                parts, width = self._draw_singleq_gate(gate_text)
+            elif gate.name == "SWAP":
+                self._update_swap_gate(gate)
+                continue
+            else:
+                merged_wire = sorted(gate.targets + (gate.controls or []))
+                wire_list = list(range(merged_wire[0], merged_wire[-1] + 1))
+                parts, width = self._draw_multiq_gate(gate, gate_text)
+
+            # update the render strings for the gate
+            layer = max(len(self._layer_list[i]) for i in wire_list)
+            xskip = self._get_xskip(wire_list, layer)
+            self._manage_layers(width, wire_list, layer, xskip)
+
+            if isinstance(gate, Measurement):
+                self._adjust_layer(wire_list, xskip)
+                self._update_singleq(gate.targets, parts)
+                self._update_cbridge(gate, wire_list, xskip, width)
```
For clarity
```python
            elif gate.name == "SWAP":
                self._update_swap_gate(gate)
                continue
```
should be moved to later part, together with other `_update_*` methods.

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+                    self._render_strs["mid_frame"][wire] += mid_bar_conn
+                    self._render_strs["bot_frame"][wire] += bar_conn
+
+    def _update_swap_gate(self, gate: Gate):
+        """
+        Update the render strings for the SWAP gate
+        """
+
+        wire_list = list(range(min(gate.targets), max(gate.targets) + 1))
+        layer = max(len(self._layer_list[i]) for i in wire_list)
+        xskip = self._get_xskip(wire_list, layer)
+        self._adjust_layer(wire_list, xskip)
```
This part could be removed once ` _update_swap_gate` is moved to later part in `self.layout`

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+
+        return (top_frame, mid_frame, bot_frame), len(top_frame)
+
+    def _draw_multiq_gate(self, gate: Gate, gate_text: str):
+        """
+        Draw a multi qubit gate
+
+        Parameters
+        ----------
+        gate : Gate
+            The gate to be drawn.
+
+        Returns
+        -------
+        tuple
+            The parts of the gate to be drawn.
+            i.e. the top_frame, mid_frame, mid_connect, mid_connect_label, and bot_frame.
+
+        int
+            The width of the gate.
+        """
+
+        lid_seg = "─" * (ceil(self.style.gate_pad) * 2 + len(gate_text))
+        pad = " " * ceil(self.style.gate_pad)
+
+        top_frame = f" ┌{lid_seg}┐ "
+        bot_frame = f" └{lid_seg}┘ "
+        mid_frame = f" │{pad}{' ' * len(gate_text)}{pad}│ "
+        mid_connect = f"─┤{pad}{' ' * len(gate_text)}{pad}├─"
+        mid_connect_label = f"─┤{pad}{gate_text}{pad}├─"
```
Low priority:
I feel like ` _draw_singleq_gate` can be merged into this `_draw_multiq_gate`.

gadhvirushiraj replied:
Sure, the `_draw_singleq_gate` function follows a similar structure to MatRenderer, with separate drawing functions for single and multi-qubit gates.

BoxiLi commented on src/qutip_qip/circuit/base_renderer.py:
```diff
@@ -0,0 +1,212 @@
+        wire_list: List[int],
+        layer: int,
+        xskip: float = 0,
+    ) -> None:
+        """
+        Manages and updates the layer widths according to the gate's width just plotted.
+
+        Parameters
+        ----------
+        gate_width : float
+            The width of the gate to be plotted.
+
+        wire_list : list
+            The list of wires the gate is acting on (control and target).
+
+        layer : int
+            The layer the gate is acting on.
+
+        xskip : float, optional
+            The horizontal value for getting to requested layer. The default is 0.
+        """
+
+        for wire in wire_list:
+            # check if requested layer exists for the wire
+            if len(self._layer_list[wire]) > layer:
+                # check if the layer width is greater than new layer width
+                if (
+                    self._layer_list[wire][layer]
+                    < gate_width + self.style.gate_margin * 2
+                ):
+                    # update with new layer width
+                    self._layer_list[wire][layer] = (
+                        gate_width + self.style.gate_margin * 2
+                    )
+            else:
+                # add layer width: new layer width + missing layer widths if exits
+                temp = xskip - sum(self._layer_list[wire]) if xskip != 0 else 0
+                self._layer_list[wire].append(
+                    temp + gate_width + self.style.gate_margin * 2
+                )
```
Personally, I think it would be nice to explicitly return the self._layer_list just to make it clear that this has been changed. But it is up to you.

BoxiLi commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+        lid_seg = "─" * (ceil(self.style.gate_pad) * 2 + len(gate_name))
+        pad = " " * ceil(self.style.gate_pad)
+
+        top_frame = f" ┌{lid_seg}┐ "
+        mid_frame = f"─┤{pad}{gate_name}{pad}├─"
+        bot_frame = f" └{lid_seg}┘ "
+
+        return (top_frame, mid_frame, bot_frame), len(top_frame)
```
Should we add an assert checking that the length of all three should be the same? What do you think? This could even replace the last two tests.

gadhvirushiraj replied:
It can replace one test, `test_parts_len`, and the other test checks if all `_render_str` or wire lengths are consistent. I'm okay with either option; the choice might come down to which provides better coverage, while both options ensure testing. Let me know what you think.

BoxiLi replied:
It won't affect the coverage because the general test will also cover these lines of code I think.

The point is test_parts_len is to find some specific error and make debugging easier. If it is equivalent to adding assert here. I think it is better to have assert here because it helps detect problems early, where the cause is clear, rather than later when some other operation fails.

gadhvirushiraj commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+            "bot_frame": ["  "] * (self._qwires + self._cwires),
+        }
+
+    def _adjust_layer(self, wire_list: List[int], xskip: int):
+        """
+        Adjust the layers by filling the empty spaces with respective characters
+        """
+
+        for wire in wire_list:
+            self._render_strs["top_frame"][wire] += " " * (
+                xskip - len(self._render_strs["top_frame"][wire])
+            )
+            self._render_strs["bot_frame"][wire] += " " * (
+                xskip - len(self._render_strs["bot_frame"][wire])
+            )
+            if wire < self._qwires:
+                self._render_strs["mid_frame"][wire] += "─" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+            else:
+                self._render_strs["mid_frame"][wire] += "═" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
```
Since we're dealing with strings and the gates vary in width, `_adjust_layer` adds padding to maintain consistent layer alignment across the different wires. The function `_add_wire_all` appears to be of different menaing, but to clarify its purpose, renaming it to `_adjust_layer_pad` would make its role clearer.

gadhvirushiraj commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+        self._qc = qc
+        self._qwires = qc.N
+        self._cwires = qc.num_cbits
+        self._layer_list = {i: [] for i in range(self._qwires + self._cwires)}
```
This was originally used to address the classical wire issue with measurement joint, though it was handled differently. I'll change it to list.

gadhvirushiraj commented on src/qutip_qip/circuit/text_renderer.py:
```diff
@@ -0,0 +1,497 @@
+                if gate.controls:
+                    sorted_controls = sorted(gate.controls)
+                    # check if the top of the gate is closer to the first control wire
+                    is_top_closer = wire_list[0] >= sorted_controls[0]
+                    closest_pos = (
+                        wire_list[0] if not is_top_closer else wire_list[-1]
+                    )
+                    self._update_qbridge(
+                        gate,
+                        list(
+                            range(
+                                min(sorted_controls[0], closest_pos),
+                                max(sorted_controls[-1], closest_pos) + 1,
+                            )
+                        ),
```
> "What actually will happen if a gate is controlled by q0 and q2 and target on q1?"

This case is not handled as of now, will add logic to tackle it.

BoxiLi replied:
For some reason it already works ;)

![image](https://github.com/user-attachments/assets/8b77ca2b-3697-41ed-b157-e8e0bb61fedd)

BoxiLi commented on doc/source/qip-basics.rst:
```diff
@@ -272,31 +272,136 @@ gate function returning a :class:`qutip.Qobj` and save it in the attribute ``use
+        # plot the quantum circuit using  MatRenderer
+        qc.draw("matplotlib")
+
+    .. image:: /figures//mat_circuit_example.png
```
It is better to use the plot derivative than upload the `png` here. We used it for latex because there is really no other way.

https://github.com/qutip/qutip-qip/blob/master/doc/source/qip-processor.rst#spinchain

It needs to be rebased to remove png from the history.

BoxiLi replied:
You can download and view the rendered `html` here in the artifact.

https://github.com/qutip/qutip-qip/actions/runs/10496547795?pr=244

BoxiLi commented on doc/source/qip-basics.rst:
```diff
@@ -272,31 +272,136 @@ gate function returning a :class:`qutip.Qobj` and save it in the attribute ``use
+
+
+    .. code-block:: none
+    
+               ┌──────┐  ┌──────┐  ┌───────┐  ┌───┐   
+        q1 :───┤ CNOT ├──┤ SNOT ├──┤       ├──┤ M ├───
+               └───┬──┘  └──────┘  │       │  └─╥─┘   
+                   │               │       │    ║     
+        q0 :───────▇───────────────┤ ISWAP ├────║─────
+                                   └───────┘    ║     
+                                                ║     
+        c0 :════════════════════════════════════╩═════
+
```
Somehow this line 367 is shifted？

BoxiLi commented on tests/test_matrenderer.py:
```diff
@@ -0,0 +1,14 @@
+import pytest
+from unittest.mock import patch
+from tests.test_textrenderer import qc1, qc2, qc3
+
+
+@pytest.mark.parametrize("qc_fixture", ["qc1", "qc2", "qc3"])
+def test_matrenderer(request, qc_fixture):
+    qc = request.getfixturevalue(qc_fixture)
+
+    with patch("matplotlib.pyplot.show"):
+        try:
+            qc.draw("matplotlib")
+        except Exception as e:
+            assert False, f"Error: {e}"
```
Let's merge the test files since it is just one test you are already import the circuits from the other one.

And there is no need to try and catch because pytest will automatically catch the exceptions

BoxiLi reviewed (commented):
Looks good!

Do you think the MatRenderer and TextRenderer and the style class should be exposed to the users? If so, the public API should be added to the apidoc https://github.com/qutip/qutip-qip/blob/master/doc/source/apidoc/qutip_qip.circuit.rst.

BoxiLi commented:
And this comments of min:


> A design question, technically speaking, the Renderer class should be independent of the quantum circuit provided.
> 
> Therefore, qc should better be an input argument for some canvas_plot for draw methods. Not as part of the initialization of the renderer class.
>

BoxiLi commented:
We can (and should) replace other latex circuits in the doc with the new matplotlib plotting. But let's not add that to this PR just for clarity.

BoxiLi commented:
Add the default value of the style configuration to the doc.

gadhvirushiraj commented:
@BoxiLi 
I've updated the `circuit.py` file. The `QubitCircuit` object now uses matplotlib as the default representation (if matplotlib is available); otherwise, it falls back to text-based rendering. I've also removed the unnecessary LaTeX hooks for Jupyter Notebooks. Currently, I've added a `bool option` to the draw function for saving in the respective format, didn't find  a better approach for integrating the save functionality.

BoxiLi commented on src/qutip_qip/circuit/circuit.py:
```diff
@@ -940,89 +947,77 @@ def compute_unitary(self):
-        file_name : Filename of the exported image.
-                    Default : "exported_pic"
+        file_path : str, optional
+            The path to save the image file. Default is the current working directory.
```
It seems that the default is `circuit` not current working directory

BoxiLi reviewed (commented):
Yeah the interface of `circuit.draw` looks good.

BoxiLi commented on src/qutip_qip/circuit/circuit.py:
```diff
@@ -940,84 +947,77 @@ def compute_unitary(self):
-        file_path : Path to which the file has to be exported.
-                    Default : ""
-                    Note : User should have write access to the location.
+        kwargs : dict
+            Additional keyword arguments passed to the renderer.
```
Need to point to the doc where all the kwargs  are listed.

gadhvirushiraj replied:
All the kwarg arguments are mentioned in `base_renderer.py`'s `StyleConfig Dataclass`, I have added a note.

BoxiLi commented on doc/source/qip-basics.rst:
```diff
@@ -272,31 +272,168 @@ gate function returning a :class:`qutip.Qobj` and save it in the attribute ``use
+
+- **Text**:
+
+  .. testcode::
```
I found the problem @gadhvirushiraj . The indent matters here! `testcode`, `testoutput` and `**Customization Parameters**` should not have indent.

BoxiLi commented:
Sorry for the delay and thank you very much for the fix @gadhvirushiraj ! It looks all good. Merge.

Lets see how the tutorials run!
